python/semantic_kernel/kernel.py

```diff
@@ -43,17 +49,22 @@ class Kernel(KernelBase):
 
     def __init__(
         self,
-        skill_collection: SkillCollectionBase,
-        prompt_template_engine: PromptTemplateEngineBase,
-        memory: SemanticTextMemoryBase,
-        config: KernelConfig,
-        log: Logger,
+        *,
+        skill_collection: Optional[SkillCollectionBase] = None,
+        prompt_template_engine: Optional[PromptTemplateEngineBase] = None,
+        memory: Optional[SemanticTextMemoryBase] = None,
+        memory_storage: Optional[MemoryStoreBase] = None,
+        config: Optional[KernelConfig] = None,
+        log: Optional[Logger] = None,
     ) -> None:
-        self._log = log
-        self._config = config
-        self._skill_collection = skill_collection
-        self._prompt_template_engine = prompt_template_engine
-        self._memory = memory
+        self._log = log if log else NullLogger()
+        self._config = config if config else KernelConfig()
+        self._skill_collection = skill_collection if skill_collection else SkillCollection(self._log)
+        self._prompt_template_engine = prompt_template_engine if prompt_template_engine else PromptTemplateEngine(self._log)
+        self._memory = memory if memory else NullMemory()
+
+        if memory_storage:
```

Comment:
this is one of those cases why the builder is a better path. With a growing set of dependencies, this approach turns the kernel ctor into a big builder logic. The builder allows to move the composition out, so for example the kernel doesn't need to know that memory internally uses memory_storage

Reply:
One suggestion here is to convert the class into a dataclass. That way, all the `x if x else NullX` can be converted to `x: XType = field(default_factory=NullX)`.
```suggestion
       @dataclass
       class Kernel(KernelBase):
           _log: Logger = field(default_factory=NullLogger)
           _config: field(default_factory=KernelConfig)
```

One downside of this is that if you want to instantiate the class and call the constructor using keyword arguments, the keyword arguments will be starting with `_`. This is one of the reasons why I prefer `attrs` to dataclasses. With `attrs` the above definition would automatically translate to keyword arguments without underscores, but the attributes will automatically be considered "private".

Reply:
Do we have to make the attributes private?

Reply:
I'm not opposed to making them public.